libkbfs/tlf_journal.go

```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
+		// the initializer, but at this point we don't know if we'll
+		// need to initialize or not.
+		ibrmds, err = j.mdJournal.getRange(jStatus.RevisionStart, stop)
+		if err != nil {
+			return TLFJournalStatus{}, nil, nil, false, err
+		}
 	}
-	return jStatus, ibrmds, complete, nil
+
+	return jStatus, ibrmds, unflushedPaths, complete, nil
 }
 
-func (j *tlfJournal) getJournalStatusWithPaths(ctx context.Context,
-	cpp chainsPathPopulator) (TLFJournalStatus, error) {
-	jStatus, ibrmds, complete, err := j.getJournalStatusWithRange()
-	if err != nil {
-		return TLFJournalStatus{}, err
-	}
+func (j *tlfJournal) batchConvertImmutables(ctx context.Context,
+	ibrmds []ImmutableBareRootMetadata) ([]ImmutableRootMetadata, error) {
+	irmds := make([]ImmutableRootMetadata, 0, len(ibrmds))
+	if len(ibrmds) > 0 {
+		ibrmdBareHandle, err := ibrmds[0].MakeBareTlfHandle(nil)
+		if err != nil {
+			return nil, err
+		}
 
-	if len(ibrmds) == 0 {
-		return jStatus, nil
-	}
+		handle, err := MakeTlfHandle(
+			ctx, ibrmdBareHandle, j.config.usernameGetter())
+		if err != nil {
+			return nil, err
+		}
 
-	ibrmdBareHandle, err := ibrmds[0].MakeBareTlfHandle(nil)
-	if err != nil {
-		return TLFJournalStatus{}, err
-	}
+		for _, ibrmd := range ibrmds {
+			irmd, err := j.convertImmutableBareRMDToIRMD(ctx, ibrmd, handle)
+			if err != nil {
+				return nil, err
+			}
 
-	irmds := make([]ImmutableRootMetadata, 0, len(ibrmds))
-	handle, err := MakeTlfHandle(
-		ctx, ibrmdBareHandle, j.config.usernameGetter())
-	if err != nil {
-		return TLFJournalStatus{}, err
+			irmds = append(irmds, irmd)
+		}
 	}
 
-	for _, ibrmd := range ibrmds {
-		irmd, err := j.convertImmutableBareRMDToIRMD(ctx, ibrmd, handle)
+	return irmds, nil
+}
+
+func (j *tlfJournal) getJournalStatusWithPaths(ctx context.Context,
+	cpp chainsPathPopulator) (jStatus TLFJournalStatus, err error) {
+	// This loop is limited only by the lifetime of `ctx`.
+	var unflushedPaths unflushedPathsMap
+	var complete bool
+	for {
+		var ibrmds []ImmutableBareRootMetadata
+		jStatus, ibrmds, unflushedPaths, complete, err =
+			j.getJournalStatusWithRange()
 		if err != nil {
 			return TLFJournalStatus{}, err
 		}
 
-		irmds = append(irmds, irmd)
-	}
-
-	// Make chains over the entire range to get the unflushed files.
-	// TODO: cache this chains object and update it when new revisions
-	// appear and when revisions are flushed, instead of rebuilding
-	// every time.
-	chains := newCRChainsEmpty()
-	for _, irmd := range irmds {
-		winfo := writerInfo{
-			uid:      j.uid,
-			kid:      j.key.KID(),
-			revision: irmd.Revision(),
-			// There won't be any conflicts, so no need for the
-			// username/devicename.
-		}
-		err := chains.addOps(j.config.Codec(), irmd.data, winfo,
-			irmd.localTimestamp)
-		if err != nil {
-			return TLFJournalStatus{}, nil
-		}
-	}
-	chains.mostRecentMD = irmds[len(irmds)-1]
+		// We need to make or initialize the unflushed paths.
+		if unflushedPaths == nil {
```

Comment:
Seems clearer (and less-indented) to do:

```
// Use the cached paths.
if unflushedPaths != nil {
  break
}

// Otherwise, figure out the paths ... but don't cach it.
if !complete {
  ...
  break
}

// We need to...
doInit, err := ...
```